### PR TITLE
Update spec to require diffSide for changes

### PR DESCRIPTION
**Pull request by drdavella**
Two notes:
* I still think that making this a _required_ field is not quite correct given that the vast majority of cases are going to be covered by `"right"`, which suggests a reasonable default
* While this does ultimately control which side of the diff is used to present this comment, that feels like a platform implementation detail. It might make more sense to call this `"changeType"` and make the values `"add"`, `"update"`, and `"remove"`. Making codemodder differentiate between `"add"` and `"update"` seems needlessly complicated, though.

### CODE REVIEW

**Comment by gilday** [comment]
> While this does ultimately control which side of the diff is used to present this comment, that feels like a platform implementation detail. It might make more sense to call this "changeType" and make the values "add", "update", and "remove". Making codemodder differentiate between "add" and "update" seems needlessly complicated, though.

This tracks closely with "the other" solution that we previously listed as a solution candidate in [CodeTF May Describe Comments on Original and Added Code](https://www.notion.so/pixee/CodeTF-May-Describe-Comments-on-Original-and-Added-Code-31667692f6cc417ea8c75b27ca2c925b#d13e56dd6ba44eca9fa683e3c2d7cff2). I felt like we had consensus around the "add `diffSide` to CodeTF" solution, so I removed it from the ticket, but if we're going to revisit that, then now is the time.

The "other solution" I'm talking about is the one wherein Pixeebot infers the diff side based on the actual diff. This solution differs from the `changeType` that you've proposed only that it assumes consumers of the CodeTF can figure out the change type themselves by looking at the diff.

* all added lines? --> additive, diff side right
* changed lines? --> modification, diff side right
* all removed lines? --> removal, diff side left

Regardless of whether codemodder explicitly tells the platform the `changeType` or the platform infers it from the diff, this is a very leaky abstraction, because codemodder must be aware of this behavior so that it can send the correct `lineNumber`. That is, it must send a `lineNumber` value corresponding to the new file when the change type is either `add` or `update`, and it must send a `lineNumber` value corresponding to the original file for the `remove` change type.

In theory, the `changeType` abstraction gives the platform flexibility to handle this data differently, but because this abstraction is leaky, that's actually not the case. Consider the scenario where the platform decides it wants to move comments for `updated` changes from the original file to the new file (from left to right). The platform cannot simply decide that comments with `changeType` == `update` should now go on the right instead of the left (or vice versa), because the `lineNumber` is specific to a side.

Because the `lineNumber` is specific to a side, I think it makes sense to include the side in CodeTF, so that there's no guessing about codemodder's intentions for what that `lineNumber` field means.

File: codetf.json [hunk 1]
```diff
@@ -37,6 +37,7 @@
                 "changes" : [
                     {
                         "lineNumber" : "153", // the "primary" line number that was changed (required)
+                        "diffSide": "right", // "right" or "left". "right" indicates a new or modified line of code. "left" indicates a line of code that was deleted (required)
```
Comment (gilday) [inline codetf.json:40]: consider renaming this `side` to align with GitHub's API (see [create PR comment endpoint docs](https://docs.github.com/en/rest/pulls/comments?apiVersion=2022-11-28#create-a-review-comment-for-a-pull-request)). I don't think "alignment with GitHub API" is a strict goal, but if there's no reason not to, we may as well, to keep our code easier to read.

**Comment by drdavella** [comment]
> Because the lineNumber is specific to a side, I think it makes sense to include the side in CodeTF, so that there's no guessing about codemodder's intentions for what that lineNumber field means.

After reading through your description, I agree. The idea about automatically identifying which side is really clever and is no doubt possible in principle. However, given that codemodder already _necessarily_ has all of the relevant information (since otherwise it couldn't have generated the change at all), it also seems needlessly complex.

That leaves the issue of whether `diffSide` should be required. I think we have differing opinions on this but mine is not too strongly held. I mostly feel like making it optional allows for backwards compatibility, and it also seems justified given that there is a reasonable default for the majority of cases.

File: codetf.json (hunk 1 above)
Comment (gilday) [inline codetf.json:40]: > "right" indicates a new of modified line of code.

This is only true for Python. This discrepancy between the two codemodders is the last issue that I don't believe we've yet built consensus towards, and I would like to.

Codemodder Python and Codemodder Java are not consistent with how they comment on modified code. Python comments use side `right` to describe how the codemod's change improves the user's code, while Java comments use side `left` to describe why the user's code needed improving.

Ultimately, I think we will find use cases for both kinds of comments, but I also believe (as I know you do @drdavella) that one kind of comment makes sense in an overwhelming majority of cases. Unlike @drd, I don't have strong opinions on which side is correct in the majority of cases when describing modified code. @nahsra can you and @drdavella discuss and come to a consensus on which `side` to use to describe modified code, so that codemod storytelling is consistent across implementations? I would love to put guidance into the CodeTF spec for the `diffSide` field e.g. 

> "right" or "left". "right" is for additive changes, while `left` describe lines removed. **Typically**, modified code should use `right` to describe why the new code is an improvement on the old code (required)

**Review by gilday — CHANGES_REQUESTED** [review]
Can we build consensus towards using `right` or `left` for modified code _most of the time_ while leaving the door open for edge cases where it makes sense to use the other side? Arriving at this consensus is the only thing I feel strongly that we need to do before we can ship this.

File: codetf.json (hunk 1 above)
Comment (drdavella) [reply codetf.json:40]: I'm a little worried that by doing that we obscure the intent of this field since understanding its purpose would mean having some context about the underlying implementation of the consumer. I'd prefer to preserve as much clarity here as possible, although I'm certainly open to a better name than `diffSide`. That's what brought me originally to the idea of `changeType`.

What about something that makes its relationship to the line number more explicit? Like `lineNumberLocation` and the values are `"original"` and `"new"`?

tl;dr: naming things is hard.

Comment (drdavella) [reply codetf.json:40]: I agree with this. I do feel reasonably strongly that comments should be phrased in terms of fixes rather than problems given the overall philosophy of codemodder. However, I don't want to preclude any flexibility here. What do you think about the proposal for something like `lineNumberLocation` in this context? It feels a little more neutral in terms of purpose and also maybe makes the rationale a little more explicit?

Comment (gilday) [reply codetf.json:40]: For this discussion, I don't think the name of the property will provide clarity. I think we just need to come to a consensus on the fundemental question:

> should codemod storytelling be centered around why the user's code needed to be improved, or should it be centered around why the new code is an improvement.

Again, I don't have strong opinions on the outcome. I'm asking that @nahsra weigh in, because I expect he does have opinions on this. And if we ultimately agree to change codemodder-java to use `side: "right"` instead of its current (implied) `left`, then @nahsra can better speak to the level of effort to make that change, which will give us an idea of how long we can expect the two to be out of sync.

Comment (drdavella) [reply codetf.json:40]: I understand, but my point is that CodeTF does not and should not know anything at all about the GitHub API. What is more compelling is to use a field that makes sense in the context of a CodeTF change. 

In the context of CodeTF what we're really trying to say is whether the given line number and description refer to the original code or whether they refer to new code. I think we should do our best to make that connection explicit when naming this field (and I admit that `diffSide` does no such thing) rather than coupling this field to a largely contingent and invisible implementation detail of our platform.

If you feel strongly that `side` is the most useful term here, then I would argue that it really belongs in the `properties` bucket in which case it can't be a required field.

Comment (drdavella) [reply codetf.json:40]: Fwiw, the change that would make me the happiest is to add `side` to the `properties` bucket since it is explicitly for such platform-specific purposes. However, I approached this with the assumption that this needs to be a required field, which necessitates a spec change.

If you're okay with using `properties` for this purpose, then I will close this PR and make the change to Python codemodder. This means we need to agre on a reasonable default. As a compromise I am happy to conform to the existing platform behavior and make Python explicit in its requirements. I think this would preclude the need for a lot of the other discussions on this thread since we can adapt to our needs without further spec changes.

The downside of this is that future codemodders are going to get the default behavior, which may not be what they want. But that's a problem for future you and me, and not necessarily one that needs to be specced.

Comment (gilday) [reply codetf.json:40]: I think the `properties` bag should be reserved only for cases where we can't possibly justify including some data in the CodeTF spec proper. It's ripe for abuse and I kind of wish it didn't exist.

We'll just map whatever term you choose to `side`.

Comment (nahsra) [reply codetf.json:40]: There's a lot here, and I only want to state a few opinions and then get out of the way:

1. I agree that I would like CodeTF to be as neutral as possible to platform or SCM needs. I think `diffSide` is more clear to codemod authors, who we want to attract. The platform is a bit of a captured audience.
2. We have 2 codemodders right now and it is convenient for them to provide opposite `diffSide` values. I think there should be no default and we should force codemodder _developers_ to figure this out for themselves. If it's required, it's easy to provide. If it's not required,  they will likely be subtly wrong and maybe not notice.
3. This should not be a `properties` entry. It is a part of storytelling, which is a top concern and differentiator for codemodder.

Comment (nahsra) [reply codetf.json:40]: My immediate reaction is: I don't recognize the problem as being real or important, we need to both, and it can be done from right or left.

I think choosing right or left can be driven off the convenience of codemod authors and their respective CST libraries. Regarding the question posed: we must do both of these things, either in the commentary or in the APIs we introduce. I might be late to this discussion and missing context.

Comment (drdavella) [reply codetf.json:40]: What about `diffSide` with `"new"` and `"original"` to make the connection more explicit?

Comment (gilday) [reply codetf.json]: > driven off the convenience of codemod authors and their respective CST libraries

In the case where both are equally as convenient, do you have some preference towards either describing the problem or describing the solution in our storytelling?

File: codetf.json [hunk 2]
```diff
@@ -38,6 +38,7 @@
                     {
                         "lineNumber" : "153", // the "primary" line number that was changed (required)
                         "description" : "Added a call to ObjectInputStream#setObjectFilter() to prevent known malicious gadgets..", // human readable description of a given change (optional)
+                    "diffSide": "right", // "right" or "left". "right" indicates that the lineNumber and description correspond to a position in the new or modified code. "left" indicates these fields correspond to a position in the original code. Descriptions of the change itself should generally use "right", except in cases where the only change is a deletion. In contrast, "left" can be used to describe the original problem being fixed. (required)
```
Comment (gilday) [inline codetf.json:41]: nit: can we adjust the indentation here

File: codetf.json (hunk 1 above)
Comment (nahsra) [reply codetf.json]: Just want to point out that these numbers could be wrong in the situation where we've made multiple changes to the same file across multiple codemods, right? The first line number will be correct, but any subsequent changes will be wrong. I think aiming for perfection here isn't possible, so I'd be nervous to keep adding complexity here.

We may eventually be able to add AI more appropriately guess the best location for these?

Comment (drdavella) [reply codetf.json]: In principle, yes. In practice, it is incumbent upon the user (e.g. the platform) to properly account for this before invoking codemodder and rendering any results.

Comment (gilday) [reply codetf.json]: Wait, why would the line numbers necessarily be wrong? Each codemod's diff builds on the previous codemod's diff, so why would this be different for the line numbers in our storytelling?

Comment (nahsra) [reply codetf.json]: Codemod A inserts a line at line 10. It records the change at line 10.

Codemod B inserts at line 5. It records the change at line 5.

Codemod A change was recorded at the correct "left" location after the next codemod executed, the recorded left" location will be wrong.

Comment (gilday) [reply codetf.json]: I see. I think it's correct to say that if the platform were commenting on _each commit's diff_ then the lines would be correct, but the platform is actually creating _PR comments_, and PRs mash-up many commits into one diff view.
